packages/core/core/src/requests/WriteBundleRequest.js

```diff
@@ -129,6 +129,16 @@ async function run({input, options, api}) {
   if (info.isLargeBlob) {
     contentStream = options.cache.getStream(cacheKeys.content);
   } else {
+    // Force a refresh of the cache to avoid a race condition
+    // between threaded reads and writes that can result in an LMDB cache miss:
+    //   1. Thread A has read some value from cache, necessitating a read transaction.
+    //   2. Concurrently, Thread B finishes a packaging request.
+    //   3. Subsequently, Thread A is tasked with this request, but fails because the read transaction is stale.
+    // This only occurs if the reading thread has a transaction that was created before the writing thread committed,
+    // and the transaction is still live when the reading thread attempts to get the written value.
+    // See https://github.com/parcel-bundler/parcel/issues/9121
+    options.cache.refresh();
```

Comment:
There's a write of the sourcemap down lower (starting ~line 174 in your changed code) - while there are a few more async calls in between here and there, and so it's more likely that LMDB will have reset the transaction by then, should we move this refresh outside of this condition so that even if we're dealing with a large blob we refresh the cache anyway to avoid (potential) failures when (maybe) getting the sourcemap from the cache?

Reply:
While thinking through this a bit more, i realized that `WriteBundleRequest` will always run in the main thread (it doesn't ever create a handle with the worker farm like the `PackageRequest` does). I think this means that we actually _can_ do something closer to the 'correct' approach by moving the reset to `WriteBundlesRequest` right after we've received the result from the `PackageRequest`.

I pushed an update that does this. The overview is now (starting at `Parcel => run(buildRequest) => run(writeBundlesRequest)`):

- [**main thread**] for each bundle:
  - if `useMainThread` is `false`
    - [**worker thread**] `run(packageRequest)`
    - [**main thread**] `cache.refresh()`
  - else
    - [**main thread**] `run(packageRequest)`
  - [**main thread**] `run(writeBundleRequest)`

Reply:
Yep, that makes sense.. I dd consider this but I had confused myself about what was running where.